README.md
# Broadband Usage Meter (OSX)

Little menu bar app that allows you to check your broadband data usage directly from the desktop. Supports Sri Lankan service providers only. Currently supports SLT.

## Installation

1. Download the latest [release]().
2. Double-click on it. The following window will appear.
3. Click the **Move to Applications Folder** button.


## Usage

## Requirements

- OS X 10.9 upwards

## Support

Check out the [roadmap] for this project to see the upcoming features. If you have any more ideas/suggestions, please open an issue or let me know via [twitter](https://twitter.com/IJNanayakkara) or [email](mailto:isuru.nan@gmail.com).

## Legal

Broadband Usage Meter is not affiliated with any of the service providers.

This application is released under MIT license. See [LICENSE](https://github.com/Isuru-Nanayakkara/Broadband-Usage-Meter-OSX/blob/master/LICENSE) for more information.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).

## [Unreleased]
### Added
- See the project [roadmap]().

## [0.0.1] - 2015-10-24
### Added
- 

Broadband Usage Meter/Broadband Usage Meter/Usage.swift
//
//  Usage.swift
//  SLT Usage Meter
//
//  Created by Isuru Nanayakkara on 10/23/15.
//  Copyright © 2015 BitInvent. All rights reserved.
//

import Foundation

struct Usage: CustomStringConvertible {
    var totalData: Double
    var remainingTotalData: Double
    var totalPeakData: Double
    var remainingPeakData: Double
    
    private let gigabyteInBytes: Double = 1073741824
    
    init?(jsonData: NSData) {
        var jsonDict = [String: AnyObject]()
        
        do {
            jsonDict = try NSJSONSerialization.JSONObjectWithData(jsonData, options: []) as! [String: AnyObject]
        } catch {
            print("Error occurred parsing data: \(error)")
            return nil
        }
        print(jsonDict)
        
        let multiplier = pow(10.0, 1)
        
        totalData = (jsonDict["totalfup"] as! NSString).doubleValue
        remainingTotalData = round((jsonDict["totalrem"] as! NSString).doubleValue / gigabyteInBytes * multiplier) / multiplier
        totalPeakData = (jsonDict["peakfup"] as! NSString).doubleValue
        remainingPeakData = round((jsonDict["peakrem"] as! NSString).doubleValue / gigabyteInBytes * multiplier) / multiplier
    }
    
    var remainingTotalDataPercentage: Double {
        let totalRemaining = 100 - totalData
        return round(totalRemaining)
    }
    
    var remainingPeakDataPercentage: Double {
        let peakRemaining = 100 - totalPeakData
        return round(peakRemaining)
    }
    
    var description: String {
        return "Total Data: \(totalData)\nRemaining Total Data: \(remainingTotalData)\nTotal Peak Data: \(totalPeakData)\nRemaining Peak Data: \(remainingPeakData)"
    }
}

Broadband Usage Meter/Broadband Usage Meter/SLTAPI.swift
//
//  SLTAPI.swift
//  SLT Usage Meter
//
//  Created by Isuru Nanayakkara on 10/23/15.
//  Copyright © 2015 BitInvent. All rights reserved.
//

import Foundation

class SLTAPI {
    
    private let session = NSURLSession.sharedSession()
    private let baseURL = "https://www.internetvas.slt.lk/SLTVasPortal-war/"
    
    func login(userID userID: String, password: String, completion: (error: NSError?) -> ()) {
        let url = NSURL(string: baseURL + "login/j_security_check/j_security_check")!
        let params = ["j_username": userID, "j_password": password]
        let request = NSMutableURLRequest(URL: url)
        request.HTTPMethod = "POST"
        request.encodeParameters(params)
        session.dataTaskWithRequest(request, completionHandler: { data, response, error in
            if let httpResponse = response as? NSHTTPURLResponse {
                if httpResponse.statusCode != 200 {
                    print("Something's fucky! - \(httpResponse)")
                    completion(error: error)
                } else {
                    let html = NSString(data: data!, encoding: NSUTF8StringEncoding) as! String
                    if html.containsString("Invalid Credentials") {
                        let errorDescription = "Invalid Credentails"
                        let recoverySuggestion = "Please enter the correct portal Username and Password"
                        let loginError = NSError(domain: "SLT-Usage-Error", code: 0, userInfo: [NSLocalizedDescriptionKey: errorDescription, NSLocalizedRecoverySuggestionErrorKey: recoverySuggestion])
                        completion(error: loginError)
                    } else {
                        completion(error: nil)
                    }
                }
            }
        }).resume()
    }
    
    func fetchUsage(completion: (usage: Usage?, error: NSError?) -> ()) {
        let url = NSURL(string: baseURL + "application/GetProfile")!
        let request = NSURLRequest(URL: url)
        session.dataTaskWithRequest(request, completionHandler: { data, response, error in
            if let error = error {
                completion(usage: nil, error: error)
            } else {
                if let usage = Usage(jsonData: data!) {
                    completion(usage: usage, error: nil)
                } else {
                    let errorDescription = "Error Fetching Data"
                    let recoverySuggestion = ""
                    let fetchDataError = NSError(domain: "SLT-Usage-Error", code: 1, userInfo: [NSLocalizedDescriptionKey: errorDescription, NSLocalizedRecoverySuggestionErrorKey: recoverySuggestion])
                    completion(usage: nil, error: fetchDataError)
                }
            }
        }).resume()
    }
}

// see: http://stackoverflow.com/a/31031859/1077789
extension NSMutableURLRequest {
    
    /// Percent escape
    ///
    /// Percent escape in conformance with W3C HTML spec:
    ///
    /// See http://www.w3.org/TR/html5/forms.html#application/x-www-form-urlencoded-encoding-algorithm
    ///
    /// - parameter string:   The string to be percent escaped.
    /// - returns:            Returns percent-escaped string.
    
    private func percentEscapeString(string: String) -> String {
        let characterSet = NSCharacterSet.alphanumericCharacterSet().mutableCopy() as! NSMutableCharacterSet
        characterSet.addCharactersInString("-._* ")
        
        return string
            .stringByAddingPercentEncodingWithAllowedCharacters(characterSet)!
            .stringByReplacingOccurrencesOfString(" ", withString: "+", options: [], range: nil)
    }
    
    /// Encode the parameters for `application/x-www-form-urlencoded` request
    ///
    /// - parameter parameters:   A dictionary of string values to be encoded in POST request
    
    func encodeParameters(parameters: [String : String]) {
        HTTPMethod = "POST"
        
        let parameterArray = parameters.map { (key, value) -> String in
            return "\(key)=\(self.percentEscapeString(value))"
        }
        
        HTTPBody = parameterArray.joinWithSeparator("&").dataUsingEncoding(NSUTF8StringEncoding)
    }
}

Broadband Usage Meter/Broadband Usage Meter/MeterView.swift
//
//  MeterView.swift
//  SLT Usage Meter
//
//  Created by Isuru Nanayakkara on 10/23/15.
//  Copyright © 2015 BitInvent. All rights reserved.
//

import Cocoa

class MeterView: NSView {
    
    @IBOutlet weak var peakLevelIndicator: NSLevelIndicator!
    @IBOutlet weak var totalLevelIndicator: NSLevelIndicator!
    @IBOutlet weak var peakPercentageLabel: NSTextField!
    @IBOutlet weak var totalPercentageLabel: NSTextField!
    
    func update(usage: Usage) {
        dispatch_async(dispatch_get_main_queue()) {
            self.peakLevelIndicator.doubleValue = usage.remainingPeakData
            self.peakLevelIndicator.toolTip = "\(usage.remainingPeakData) GB Left"
            self.peakPercentageLabel.stringValue = "\(Int(usage.remainingPeakDataPercentage))%"
            
            self.totalLevelIndicator.doubleValue = usage.remainingTotalData
            self.totalLevelIndicator.toolTip = "\(usage.remainingTotalData) GB Left"
            self.totalPercentageLabel.stringValue = "\(Int(usage.remainingTotalDataPercentage))%"
        }
    }
}


Broadband Usage Meter/Broadband Usage Meter/MeterMenuController.swift
//
//  MeterMenuController.swift
//  SLT Usage Meter
//
//  Created by Isuru Nanayakkara on 10/23/15.
//  Copyright © 2015 BitInvent. All rights reserved.
//

import Cocoa

class MeterMenuController: NSObject {
    
    @IBOutlet weak var menu: NSMenu!
    @IBOutlet weak var meterView: MeterView!
    
    let statusItem = NSStatusBar.systemStatusBar().statusItemWithLength(NSSquareStatusItemLength)
    let api = SLTAPI()
    
    var meterMenuItem: NSMenuItem!
    var preferencesWindow: PreferencesWindow!
    
    override func awakeFromNib() {
        statusItem.image = NSImage(named: "statusIcon")
        statusItem.menu = menu
        
        meterMenuItem = menu.itemWithTitle("Meters")
        meterMenuItem.view = meterView
        
        preferencesWindow = PreferencesWindow()
        preferencesWindow.delegate = self
        
        login()
    }
    
    // HC2185334
    // Netass1st
    func login() {
        let userDefaults = NSUserDefaults.standardUserDefaults()
        let userID = userDefaults.stringForKey("userID")
        let password = userDefaults.stringForKey("password")
        
        if let userID = userID, let password = password {
            if userID.isEmpty || password.isEmpty {
                preferencesWindow.showWindow(nil)
            } else {
                api.login(userID: userID, password: password, completion: { error in
                    if let error = error {
                        print("Error occurred when loggin in: \(error.localizedDescription)")
                        dispatch_async(dispatch_get_main_queue(), {
                            NSAlert(error: error).runModal()
                        })
                    } else {
                        print("Successfully logged in!")
                        self.updateUsage()
                    }
                })
            }
        } else {
            preferencesWindow.showWindow(nil)
        }
    }
    
    func updateUsage() {
        api.fetchUsage { usage, error in
            if let usage = usage {
                print(usage.description)
                print("Remaning data: \(usage.remainingTotalDataPercentage)%")
                print("Remaining Peak data: \(usage.remainingPeakDataPercentage)%")
                self.meterView.update(usage)
                
                self.registerForBackgroundUpdating()
            }
        }
    }
    
    func registerForBackgroundUpdating() {
        let timer = NSTimer(timeInterval: 300, target: self, selector: Selector("updateUsage"), userInfo: nil, repeats: true)
        NSRunLoop.mainRunLoop().addTimer(timer, forMode: NSRunLoopCommonModes)
    }
    
    // MARK: - Actions
    @IBAction func quitItemClicked(sender: NSMenuItem) {
        NSApplication.sharedApplication().terminate(self)
    }
    
    @IBAction func preferencesItemClicked(sender: NSMenuItem) {
        preferencesWindow.showWindow(nil)
    }
}

// MARK: - PreferencesDelegate
extension MeterMenuController: PreferencesDelegate {
    func preferencesDidUpdate() {
        login()
    }
}


Broadband Usage Meter/Broadband Usage Meter/NSAttributedString+Extensions.swift
//
//  NSAttributedString+Extensions.swift
//  SLT Usage Meter
//
//  Created by Isuru Nanayakkara on 10/23/15.
//  Copyright © 2015 BitInvent. All rights reserved.
//

import AppKit

extension NSAttributedString {
    class func hyperlinkFromString(aString: String, withURL url: NSURL) -> NSAttributedString {
        let attrString = NSMutableAttributedString(string: aString)
        let range = NSMakeRange(0, attrString.length)
        
        attrString.beginEditing()
        attrString.addAttribute(NSLinkAttributeName, value: url.absoluteString, range: range)
        attrString.addAttribute(NSForegroundColorAttributeName, value: NSColor.blueColor(), range: range)
        attrString.addAttribute(NSUnderlineStyleAttributeName, value: NSNumber(integer: NSUnderlineStyle.StyleSingle.rawValue), range: range)
        attrString.endEditing()
        
        return attrString
    }
}

Broadband Usage Meter/Broadband Usage Meter/PreferencesWindow.swift
//
//  PreferencesWindow.swift
//  SLT Usage Meter
//
//  Created by Isuru Nanayakkara on 10/23/15.
//  Copyright © 2015 BitInvent. All rights reserved.
//

import Cocoa

protocol PreferencesDelegate {
    func preferencesDidUpdate()
}

class PreferencesWindow: NSWindowController, NSWindowDelegate {

    @IBOutlet weak var userIDTextField: NSTextField!
    @IBOutlet weak var passwordTextField: NSSecureTextField!
    
    var delegate: PreferencesDelegate?
    
    override var windowNibName: String {
        return "PreferencesWindow"
    }
    
    override func windowDidLoad() {
        super.windowDidLoad()
        
        window?.center()
        window?.makeKeyAndOrderFront(nil)
        NSApp.activateIgnoringOtherApps(true)
        
        // TODO: Refector NSUserDefaults code duplication
        let userDefaults = NSUserDefaults.standardUserDefaults()
        let userID = userDefaults.stringForKey("userID")
        let password = userDefaults.stringForKey("password")
        
        if let userID = userID, let password = password {
            userIDTextField.stringValue = userID
            passwordTextField.stringValue = password
        }
        
        // TODO: Add this in the next version. Current placement of the label is ugly
//        registerLabel.allowsEditingTextAttributes = true
//        registerLabel.selectable = true
//        
//        let url = NSURL(string: "https://www.internetvas.slt.lk/SLTVasPortal-war/register/register.jsp")!
//        let str = NSMutableAttributedString(string: "If you don't have a portal account, create one ")
//        str.appendAttributedString(NSAttributedString.hyperlinkFromString("here", withURL: url))
//        registerLabel.attributedStringValue = str
//        registerLabel.alignment = .Center
    }
    
    func windowWillClose(notification: NSNotification) {
        let userDefaults = NSUserDefaults.standardUserDefaults()
        userDefaults.setValue(userIDTextField.stringValue, forKey: "userID")
        userDefaults.setValue(passwordTextField.stringValue, forKey: "password")
        
        delegate?.preferencesDidUpdate()
    }
    
}
